File: statistik-service/src/main/java/se/inera/statistics/service/warehouse/Sjukfall.java
/**
 * Copyright (C) 2017 Inera AB (http://www.inera.se)
 *
 * This file is part of statistik (https://github.com/sklintyg/statistik).
 *
 * statistik is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * statistik is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package se.inera.statistics.service.warehouse;

import se.inera.statistics.service.report.model.Kon;
import se.inera.statistics.service.report.util.Icd10RangeType;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

// CHECKSTYLE:OFF FinalClass
public class Sjukfall {
    // CHECKSTYLE:ON FinalClass
    public static final int MAX_GAP = 5;

    private int start;
    private int lan;
    private int end;
    private int kon;
    private int alder;
    private List<Integer> sjukskrivningsgrader = new ArrayList<>();
    private int sjukskrivningsgrad;
    private Set<Lakare> lakare = new HashSet<>();
    private Lakare lastLakare;
    private int[] enhets;
    private int realDays;
    private long firstIntygId;
    private List<Diagnos> diagnoses = new ArrayList<>();
    private Diagnos diagnos;
    private boolean enkelt;
    private int lastEnhet;

    private Sjukfall() {
    }

    public static Sjukfall create(SjukfallExtended extendedSjukfall) {
        final Sjukfall sjukfall = new Sjukfall();
        sjukfall.start = extendedSjukfall.getStart();
        sjukfall.end = extendedSjukfall.getEnd();
        sjukfall.diagnoses = toDiagnoses(extendedSjukfall.getAllDxs());
        sjukfall.diagnos = toDiagnos(extendedSjukfall.getLastDiagnosis());
        sjukfall.realDays = extendedSjukfall.getRealDays();
        sjukfall.kon = extendedSjukfall.getKonInt();
        sjukfall.alder = extendedSjukfall.getAlder();
        sjukfall.sjukskrivningsgrad = extendedSjukfall.getSjukskrivningsgrad();
        sjukfall.sjukskrivningsgrader = new ArrayList<>(extendedSjukfall.getSjukskrivningsgrads());
        sjukfall.lan = extendedSjukfall.getLan();
        sjukfall.lakare = extendedSjukfall.getLakare();
        sjukfall.lastLakare = extendedSjukfall.getLastLakare();
        sjukfall.enhets = toArray(extendedSjukfall.getEnhets());
        sjukfall.lastEnhet = extendedSjukfall.getLastEnhet();
        sjukfall.enkelt = extendedSjukfall.isEnkelt();
        sjukfall.firstIntygId = extendedSjukfall.getFirstIntygId();
        return sjukfall;
    }

    private static int[] toArray(Set<Integer> integers) {
        final int[] ints = new int[integers.size()];
        final ArrayList<Integer> integersList = new ArrayList<>(integers);
        for (int i = 0; i < integersList.size(); i++) {
            int nextInt = integersList.get(i);
            ints[i] = nextInt;
        }
        return ints;
    }

    private static List<Diagnos> toDiagnoses(List<SjukfallExtended.Diagnos> dxs) {
        final ArrayList<Diagnos> newDxs = new ArrayList<>();
        for (SjukfallExtended.Diagnos dx : dxs) {
            newDxs.add(toDiagnos(dx));
        }
        return newDxs;
    }

    private static Diagnos toDiagnos(SjukfallExtended.Diagnos dx) {
        return new Diagnos(dx.getDiagnoskapitel(), dx.getDiagnosavsnitt(), dx.getDiagnoskategori(), dx.getDiagnoskod());
    }

    public Kon getKon() {
        return Kon.byNumberRepresentation(kon);
    }

    @Override
    public String toString() {
        return "Sjukfall{"
                + "start=" + start
                + ", end=" + end
                + ", realDays=" + getRealDays()
                + '}';
    }

    public int getDiagnoskategori() {
        return diagnos.diagnoskategori;
    }

    public int getDiagnoskod() {
        return diagnos.diagnoskod;
    }

    public List<Integer> getIcd10CodeForTypes(List<Icd10RangeType> rangeTypes) {
        List<Integer> result = new ArrayList<>();
        for (Icd10RangeType rangeType : rangeTypes) {
            switch (rangeType) {
                case KAPITEL:
                    result.add(getDiagnoskapitel());
                    break;
                case AVSNITT:
                    result.add(getDiagnosavsnitt());
                    break;
                case KATEGORI:
                    result.add(getDiagnoskategori());
                    break;
                case KOD:
                    result.add(getDiagnoskod());
                    break;
                default: throw new UnknownRangeTypeException("Unknown range type: " + rangeTypes);
            }
        }
        return result;
    }

    public List<Integer> getAllIcd10OfTypes(List<Icd10RangeType> icd10RangeTypes) {
        List<Integer> result = new ArrayList<>();
        for (Icd10RangeType icd10RangeType : icd10RangeTypes) {
            for (Diagnos diagnose : diagnoses) {
                switch (icd10RangeType) {
                    case KATEGORI:
                        result.add(diagnose.diagnoskategori);
                        break;
                    case AVSNITT:
                        result.add(diagnose.diagnosavsnitt);
                        break;
                    case KAPITEL:
                        result.add(diagnose.diagnoskapitel);
                        break;
                    case KOD:
                        result.add(diagnose.diagnoskod);
                        break;
                    default: throw new UnknownRangeTypeException("Unknown icd range type: " + icd10RangeType);
                }
            }
        }
        return result;
    }

    public int getAlder() {
        return alder;
    }

    public int getRealDays() {
        return realDays;
    }

    public int getStart() {
        return start;
    }

    public int getEnd() {
        return end;
    }

    public int getDiagnoskapitel() {
        return diagnos.diagnoskapitel;
    }

    public Stream<Integer> getDiagnoskapitels() {
        return diagnoses.stream().map(d -> d.diagnoskapitel);
    }

    public int getSjukskrivningsgrad() {
        return sjukskrivningsgrad;
    }

    public List<Integer> getSjukskrivningsgrader() {
        return sjukskrivningsgrader;
    }

    public int getDiagnosavsnitt() {
        return diagnos.diagnosavsnitt;
    }

    public String getLanskod() {
        return String.format("%1$02d", lan);
    }

    public Set<Lakare> getLakare() {
        return lakare;
    }

    public Lakare getLastLakare() {
        return lastLakare;
    }

    public int[] getEnhets() {
        return enhets;
    }

    public int getLastEnhet() {
        return lastEnhet;
    }

    public boolean isEnkelt() {
        return enkelt;
    }

    public long getFirstIntygId() {
        return firstIntygId;
    }

    private static final class Diagnos {
        private final int diagnoskapitel;
        private final int diagnosavsnitt;
        private final int diagnoskategori;
        private final int diagnoskod;

        private Diagnos(int diagnoskapitel, int diagnosavsnitt, int diagnoskategori, int diagnoskod) {
            this.diagnoskapitel = diagnoskapitel;
            this.diagnosavsnitt = diagnosavsnitt;
            this.diagnoskategori = diagnoskategori;
            this.diagnoskod = diagnoskod;
        }
    }

    private static class UnknownRangeTypeException extends RuntimeException {

        UnknownRangeTypeException(String s) {
            super(s);
        }
    }
}

File: tools/report/lib/sjukfall.py
#!/usr/bin/env python
# -*- mode: Python; fill-column: 75; comment-column: 70; coding: utf-8 ;-*-
#
# Copyright (C) 2016 Inera AB (http://www.inera.se)
#
# This file is part of statistik (https://github.com/sklintyg/statistik).
#
# statistik is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# statistik is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# -------------------------------------------------------------------------------

from datetime import datetime, timedelta
from interval import Interval

class Sjukfall:
    alpha_time = datetime(2000, 1, 1)

    def __init__(self, intyg):
        self.intyg = [intyg]
        self.start = intyg.start
        self.slut = intyg.slut

    def __str__(self):
        s = ''
        for i in self.intyg:
            s = "{0}:{1}".format(s,i)
        return i

    def sjukgrad(self, start, slut):
        """ Get the sjukgrad from the 'last' intyg
        """
        last = self.last_intyg(start, slut)
        return last.sjukgrad

    def last_intyg(self, start, slut):
        """ Get the last (latest) intyg
        """
        last = None
        for i in self.intyg:
            if i.valid(start, slut):
                if last is None:
                    last = i
                elif i.isnewer(last):
                    last = i
        assert(last != None)
        return last

    def kon(self):
        return self.intyg[0].kon

    def match_diagnos(self, diagnos, start, slut):
        for i in self.intyg:
            if i.valid(start, slut) and i.diagnoskategori == diagnos:
                return True
        return False

    def update(self, intyg):
        self.intyg.append(intyg)
        if intyg.start < self.start:
            self.start = intyg.start
        if intyg.slut > self.slut:
            self.slut = intyg.slut
        return True

    def befattningar(self, start, slut):
        """ Get all lakarbefattningar where intyg is in the period 
        """
        lakare = {}
        
        for i in self.intyg:
            if i.valid(start,slut):
                if i.lakareid not in lakare:
                    lakare[i.lakareid] = i
                else:
                    intyg = lakare[i.lakareid]
                    if i.isnewer(intyg):
                        lakare[i.lakareid] = i
        res = {}
        for k,v in lakare.items():
            l = []
            for b in v.befattning.split(','):
                l.append(b.strip())
            res[k] = l

        assert(len(res) > 0)

        return res

    def valid(self, start, slut):
        """ Check if this sjukfall is within the interval
        """
        if self.slut < start:
            return False
        if self.start > slut:
            return False

        return True

    def age(self, start, slut):
        """ Välj den högsta åldern från sjukfall, enligt K30. Notera
            att enligt K50 så väljs åldern utifrån hela sjukfallet
        """
        alder = None
        for i in self.intyg:
            #if i.valid(start, slut):
            if True:
                if alder is None:
                    alder = i.alder
                elif i.alder > alder:
                    alder = i.alder

        assert(alder != None)

        return alder

    def match(self, intyg):
        # Check if intyg belongs in this group
        for i in self.intyg:
            if intyg.start >= i.start - 6 and intyg.start <= i.slut + 6:
                return self.update(intyg)
            if intyg.slut >= i.start - 6 and intyg.slut <= i.slut + 6:
                return self.update(intyg)
            if intyg.start < i.start and intyg.slut > i.slut:
                return self.update(intyg)

        return False

    def diagnos(self, start,slut):
        last = self.last_intyg(start, slut)
        if last:
            return last.diagnos

        return None

    def diagnosis(self, start, slut):
        """ Get all diagnosis for the period """
        res = {}
        for i in self.intyg:
            if i.valid(start, slut):
                res[i.diagnos] = 1

        return res.keys()

    def get_intervals(self):
        """ Calculate number of sjukfall days for the given period
        """
        intyg = [ i for i in self.intyg ]
        intervals = []
        # Split the intyg in intervals
        while len(intyg) > 0:
            interval = Interval(intyg.pop())
            intervals.append(interval)
            match = True
            # Keep matching until no matches
            while match:
                match = False
                tmp = []
                for i in intyg:
                    if interval.match(i):
                        match = True
                    else:
                        tmp.append(i)
    
                intyg = tmp
        return intervals

    def get_enhet_intervals(self, enhet):
        """ Calculate number of sjukfall days for the given period
        """
        intyg = [ i for i in self.intyg if i.enhet == enhet ]
        #for i in intyg:
        #   print i.enhet
        intervals = []
        # Split the intyg in intervals
        while len(intyg) > 0:
            interval = Interval(intyg.pop())
            intervals.append(interval)
            match = True
            # Keep matching until no matches
            while match:
                match = False
                tmp = []
                for i in intyg:
                    if interval.match(i):
                        match = True
                    else:
                        tmp.append(i)
    
                intyg = tmp
        return intervals


    def days(self, start, slut):
        """ Enligt krav K50 ska hela sjukfallets längd räknas in
        """
        intervals = self.get_intervals()
        tot = 0
        for i in intervals:
            tot += i.days()
        return tot

    def days_in_period(self, periods, enhet=None):
        result = []
        index  = 0
        if enhet:           
            intervals = self.get_enhet_intervals(enhet)
        else:
            intervals = self.get_intervals()
        for period in periods:
            days = 0
            for interval in intervals:
                days += interval.days_in_period(period[0], period[1])
            result.append((days, period[2]))
        return result

    def lan(self, start, slut):
        lan = None
        intyg = None
        for i in self.intyg:
            if i.valid(start, slut):
                if lan is None or i.start > intyg.start:
                    lan = i.lan
                    intyg = i
                elif i.start == intyg.start and lan != i.lan:
                    print "Collision", lan,i.lan,i.enhet,intyg.enhet
                    #assert(i.enhet != intyg.enhet)
        assert(lan != None)
        return lan

    def lakare_alderkon(self, start, slut):
        """ Get all unqiue lakare
        """
        res = {}
        for i in self.intyg:
            if not i.valid(start, slut):
                continue
            
            if i.lakareid not in res:
                res[i.lakareid] = (i.lakarkon, i.lakaralder, i.start)
            else:
                tup = res[i.lakareid]
                starttime = tup[2]
                if i.start > starttime:
                    res[i.lakareid] = (i.lakarkon, i.lakaralder, i.start)
            
        return res

    def lakare(self, enhet, start, slut):
        res = []
        for i in self.intyg:    
            assert(len(i.lakareid) > 0)
            assert(i.lakareid != None)
            if i.valid(start, slut) and i.enhet == enhet:
                if i.lakareid not in res:
                    res.append(i.lakareid)
        return res

    def match_enhet(self, start, slut, enhet):
        for i in self.intyg:
            if i.valid(start, slut) and i.enhet == enhet:
                return True
        return False

    def enheter(self, start, slut):
        res = []
        for i in self.intyg:
            if i.valid(start, slut):
                if i.enhet not in res:
                    res.append(i.enhet) 
        return res

    def over_90(self, periods):
        result = []
        intervals = self.get_intervals()
        intervals.sort(key=lambda i : i.start)
        days = 0
        tot  = 0
        for interval in intervals:
            days = interval.days()
            test = 0
            prev = 0
            interval.intyg.sort(key=lambda i : i.start)
            # Make som basic checks that the periods add up
            for i in interval.intyg:
                test += i.days()
                assert (prev < i.start)
            assert(test == interval.days())

            for i in interval.intyg:
                tot += i.days()
                if tot > 90:
                    # Include the sjukfall from the intygs
                    # start to the sjukfalls slut
                    for p in periods:
                        if self.slut < p[0]:
                            continue
                        if i.intyg_start > p[1]:
                            continue

                        result.append(p[2])
                    return result
        return []   
